internal/distribute/rest/distribute.go

```diff
@@ -0,0 +1,146 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package rest provides support for pushing witnessed checkpoints to a
+// RESTful API.
+package rest
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/golang/glog"
+	"github.com/transparency-dev/formats/log"
+	"github.com/transparency-dev/witness/internal/config"
+	i_note "github.com/transparency-dev/witness/internal/note"
+	"golang.org/x/mod/sumdb/note"
+)
+
+const (
+	// HTTPCheckpointByWitness is the path of the URL to the latest checkpoint
+	// for a given log by a given witness. This can take PUT requests to update
+	// the latest checkpoint.
+	//  * first position is for the logID (an alphanumeric string)
+	//  * second position is the witness short name (alpha string)
+	HTTPCheckpointByWitness = "/distributor/v0/logs/%s/byWitness/%s/checkpoint"
+)
+
+// Witness describes the operations the feeder needs to interact with a witness.
+type Witness interface {
+	// GetLatestCheckpoint returns the latest checkpoint the witness holds for the given logID.
+	// Must return os.ErrNotExists if the logID is known, but it has no checkpoint for that log.
+	GetLatestCheckpoint(ctx context.Context, logID string) ([]byte, error)
+}
+
+// logAndVerifier represents a log known to the distributor.
+type logAndVerifier struct {
+	config config.Log
+	sigV   note.Verifier
+}
+
+func NewDistributor(baseURL string, client *http.Client, logs []config.Log, witSigV note.Verifier, wit Witness) (*Distributor, error) {
+	lvs := make([]logAndVerifier, len(logs), len(logs))
+	for i, l := range logs {
+		logSigV, err := i_note.NewVerifier(l.PublicKeyType, l.PublicKey)
+		if err != nil {
+			return nil, err
+		}
+		lvs[i] = logAndVerifier{
+			config: l,
+			sigV:   logSigV,
+		}
+	}
+	return &Distributor{
+		baseURL: baseURL,
+		client:  client,
+		logs:    lvs,
+		witSigV: witSigV,
+		witness: wit,
+	}, nil
+}
+
+type Distributor struct {
+	// URL is the base URL of the distributor.
+	// This must support a URL of the form #HTTPCheckpointByWitness.
+	baseURL string
+	client  *http.Client
+	// Logs identifies the list of source logs whose checkpoints are being distributed.
+	logs []logAndVerifier
+
+	// WitSigV can verify the cosignatures from the witness we're distributing from.
+	witSigV note.Verifier
+	// Witness is a client for talking to the witness we're distributing from.
+	witness Witness
+}
+
+// DistributeOnce a) polls the witness b) updates the fork c) proposes a PR if needed.
```

Comment:
I think the comment needs updating (GH related bits)

Reply:
Ta.